skopt/trees.py

```diff
@@ -0,0 +1,127 @@
+"""Tree based minimization algorithms"""
+
+import numpy as np
+
+from scipy import stats
+from scipy.optimize import fmin_l_bfgs_b
+from scipy.optimize import OptimizeResult
+
+from sklearn.base import clone
+from sklearn.utils import check_random_state
+
+from .gbt import GradientBoostingQuantileRegressor
+from .utils import extract_bounds
+
+
+def _expected_improvement(x, surrogate, best_y):
+    """Evaluate expected improvement for `surrogate` model at `x`"""
+    x = x.reshape(1, -1)
+    low, mu, high = surrogate.predict(x)
+    std = (high - low) / 2.
+
+    improvement = best_y - mu
+    if std > 0.:
+        exploit = improvement * stats.norm.cdf(improvement / std)
+        explore = std * stats.norm.pdf(improvement / std)
+
+    else:
+        # CDF of a very big number is 1
+        exploit = improvement * 1
+        explore = 0.
+
+    return exploit + explore
+
+
+def _random_point(lower, upper, random_state=None):
+    """Sample a random point"""
+    rng = check_random_state(random_state)
+
+    num_params = len(lower)
+    return rng.rand(num_params) * (lower + (upper - lower))
+
+
+def tree_minimize(func, bounds, base_estimator=None, maxiter=100,
+                  random_state=None):
+    """Sequential optimisation using trees.
+
+    Parameters
+    ----------
+    func: callable
+        Function to minimize. Should take an array of parameters and
+        return the function value.
+
+    bounds: array-like, shape (n_parameters, 2)
+        ``bounds[i][0]`` should give the lower bound of each parameter and
+        ``bounds[i][1]`` should give the upper bound of each parameter.
+
+    base_estimator: a GradientBoostingQuantileRegressor
```

Comment:
Is gradient boosting strictly a tree based approach?
We just fit the negative gradient to the data in every iteration right.
I would mention gradient boosting somewhere in the name of the function.